### ISSUE TITLE
Thunderbird code signature fails

### ISSUE THREAD

**Issue by Mischosch** [autopkg/recipes#258]
The `Mozilla/Thunderbird.munki.recipe` fails with the following error:

```
Code signature verification failed. Note that all verifications can be disabled by setting the variable DISABLE_CODE_SIGNATURE_VERIFICATION to a non-empty value. Use of undefined key in variable substitution: u'USER_AGENT'
```

Used that recipe for all all last 50 versions of Thunderbird without problem. Seems like v60 of Thunderbird changed something? 

Checked the signature with `codesign --verify --verbose Thunderbird.app` in the downloaded dmg. I compared signature requirements (`codesign --display -r- Thunderbird.app`) with the settings inside `Thunderbird.download.recipe`. That matches, too.

I'm not sure, where too look further to work on this. Any hints would be appreciated.

**Comment by hjuutilainen** [comment]
Yes, Thunderbird code signature is currently broken in one of its subcomponents. The code signature does pass the verification when using codesign defaults but fails when using the `--deep` flag (which is what AutoPkg does by default and which verifies all subcomponents too):

```
$ codesign --verify --verbose /Volumes/Thunderbird/Thunderbird.app/
/Volumes/Thunderbird/Thunderbird.app/: valid on disk
/Volumes/Thunderbird/Thunderbird.app/: satisfies its Designated Requirement
```

```
$ codesign --verify --verbose --deep /Volumes/Thunderbird/Thunderbird.app/
/Volumes/Thunderbird/Thunderbird.app/: a sealed resource is missing or invalid
In subcomponent: /Volumes/Thunderbird/Thunderbird.app/Contents/Library/Spotlight/thunderbird.mdimporter
file added: /Volumes/Thunderbird/Thunderbird.app/Contents/Library/Spotlight/thunderbird.mdimporter/Contents/._Info.plist
file added: /Volumes/Thunderbird/Thunderbird.app/Contents/Library/Spotlight/thunderbird.mdimporter/Contents/Resources/._schema.xml
file added: /Volumes/Thunderbird/Thunderbird.app/Contents/Library/Spotlight/thunderbird.mdimporter/Contents/Resources/English.lproj/._InfoPlist.strings
file added: /Volumes/Thunderbird/Thunderbird.app/Contents/Library/Spotlight/thunderbird.mdimporter/Contents/Resources/English.lproj/._schema.strings
```

I've reported this to Mozilla (https://bugzilla.mozilla.org/show_bug.cgi?id=1482788), let's see if they are going to fix the signature.

**Comment by aysiu** [comment]
So, just to clarify, this isn't something that can be fixed by updating the recipe. Mozilla has to fix it?

**Comment by gregneagle** [comment]
Depends on your definition of "fix", and whether of not Mozilla agrees their code signing is broken.

You can certainly work around the issue for now by turning off code signature verification in a recipe override.

**Comment by fseesink** [comment]
I tried that, turning off code signature verification in my recipe override of

`AutoPkg/RecipeRepos/com.github.autopkg.recipes/Mozilla/Thunderbird.munki.recipe`

However, while it did download a newer version than what I had (**v52.9.1**), it only downloaded **v60.0**.

I notice that if you look in the file **MozillaURLProvider.py**, you see the URL pattern used is

`MOZ_BASE_URL = "https://download.mozilla.org/?product=%s-%s&os=osx&lang=%s"`

And if you fill that in and put that in your browser of choice, e.g.,

`https://download.mozilla.org/?product=thunderbird-latest&os=osx&lang=en-US`

what you download is, in fact, as of when I type this on Monday, 8 Oct 2018,

> Thunderbird 60.0.dmg

And if you download that file, mount it, and look at the binary, you see it is **v60.0** (which came out on 6 Aug 2018).  Now that's odd, considering the latest version is **v60.2.1** (released 2 Oct 2018), and if you visit this page:

`https://www.thunderbird.net/en-US/thunderbird/all/`

and then manually choose to download the macOS English (US) file, the link you get is

`https://download.mozilla.org/?product=thunderbird-60.2.1-SSL&os=osx&lang=en-US`

I don't recall the "**-SSL**" bit the last time I went under the hood to look into this, but that's been awhile, so maybe this isn't anything new.  But it caught my eye.

Anyway, right now turning off the code signature verification does not, in fact, "work around the issue."

**Comment by fseesink** [comment]
Hmmm.  I've always had a bookmark to
`https://www.mozilla.org/en-US/thunderbird/`

but I now notice it redirects to
`https://www.thunderbird.net/en-US/` (which fits as I'm in the US)

and if I Google, the latter is the URL that comes up.  Not sure when that happened.  But wasn't looking, so maybe it's been that way a long time.

Is it possible the URL structure used in the Python script is no longer being maintained at Mozilla.org, as in if you go to
`https://download.mozilla.org/?product=thunderbird-latest&os=osx&lang=en-US`
that **thunderbird-latest** isn't being updated to point at the latest?  And if so, is this because someone isn't paying attention, or they consciously chose to no longer do that?  And how would we find out, or who would we contact?

**Comment by fseesink** [comment]
Ok, first, sorry for muddying the waters here.  This issue was initially about code signature failure, and I kind of glommed onto that with the URL issue not providing the latest Thunderbird release.

But for what it's worth, there's a bug ticket over on TB's BugZilla about the download URL not working:

https://bugzilla.mozilla.org/show_bug.cgi?id=1497574

And note one of the last posts talks about how they have a URL that provides you with JSON data showing the latest versions of Thunderbird here:
`https://product-details.mozilla.org/1.0/thunderbird_versions.json`
Output:
> {"LATEST_THUNDERBIRD_ALPHA_VERSION": "54.0a2", "LATEST_THUNDERBIRD_VERSION": "60.2.1", "LATEST_THUNDERBIRD_DEVEL_VERSION": "63.0b1", "LATEST_THUNDERBIRD_NIGHTLY_VERSION": "64.0a1"}

and that from that you can parse out the LATEST_THUNDERBIRD_VERSION and then construct a proper URL to get the latest version:
`https://download.mozilla.org/?product=thunderbird-60.2.1-SSL&os=osx&lang=en-US`

Would it make sense to adjust the MozillaURLProvider.py script to make use of this?
I played around a bit and found they do, in fact, have a similar URL with JSON output for Firefox:
https://product-details.mozilla.org/1.0/firefox_versions.json

Anyway, just thought I'd throw this out there.

**Comment by fseesink** [comment]
Ok, it's been 19 days since my last post, and no response from anyone, either here or over at bugzilla.mozilla.org.  Since what I mentioned here regarding the URL no longer working is really another issue, will create a new issue for that with hopes someone takes a look.

**Comment by apizz** [comment]
Can this be closed?

**Comment by nmcspadden** [comment]
Does the recipe currently work, or not?
